inc/saitypes.h

```diff
@@ -1355,19 +1355,49 @@ typedef sai_status_t (*sai_bulk_object_get_attribute_fn)(
         _In_ sai_bulk_op_error_mode_t mode,
         _Out_ sai_status_t *object_statuses);
 
+/**
+ * @brief SAI statistics modes
+ *
+ * Used in get statistics extended or query statistics capabilities
+ * Note enum values must be powers of 2 to be used as bit mask for query statistics capabilities
+ *
+ * @flags Contains flags
+ */
 typedef enum _sai_stats_mode_t
 {
     /**
      * @brief Read statistics
      */
-    SAI_STATS_MODE_READ,
+    SAI_STATS_MODE_READ = 1 << 0,
 
     /**
      * @brief Read and clear after reading
      */
-    SAI_STATS_MODE_READ_AND_CLEAR,
+    SAI_STATS_MODE_READ_AND_CLEAR = 1 << 1,
 } sai_stats_mode_t;
 
+typedef struct _sai_stat_capability_t
+{
+    /** Stat enum value */
+    sai_stat_id_t stat_enum;
+
+    /**
+     * @brief Bit mask of supported statistics modes (sai_stats_mode_t)
+     *
+     * For example, if read and read_and_clear are supported, value is
+     * SAI_STATS_MODE_READ | SAI_STATS_MODE_READ_AND_CLEAR
+     */
+    uint64_t stat_modes;
```

Comment:
Right now the biggest statistics enum is ~150 for port
So we are talking about allocation of 150 * (4 + 8) = ~1.8KB with uint64, vs 150 * (4 + 4) = ~1.2KB with uint32
While I agree we probably won't have more than few (maybe 5?) modes, since perhaps there is a scenario we don't anticipate now, and trying to give maximum flexibility for future growth, while paying relatively small price (extra 0.5 KB allocation), I think 64 bits are preferable. Also, in terms of usage, uint32 and uint64 are the same, as both atomic types so it doesn't complex things

Reply:
considering the alignment, since sai_stat_id_t is 32 bits, if we can make stat_modes 32 bits, then the structure will be packed into 64-bit structure. if stat_mode is 64 bits, then it could be 128 bits depending on the structure alignment.

wonder if we can use 32-bit for now and extend to 64-bit if needed in the future. that is still backward compatible?

Reply:
the function signature won't change if we modify in the future from uint32 to uint64. some code checking the modes capability may change, or may not. in any case, this is very minor change, and since it seems for the moment uint32 is prefered, i changed to using it